File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeRAGMetrics.kt
package io.github.rajibsarwar.edgeframeworks

data class EdgeRetrievalMetrics(
    val embeddingMilliseconds: Double,
    val searchMilliseconds: Double,
    val totalMilliseconds: Double,
    val resultCount: Int,
    val topScore: Float?,
    val bottomScore: Float?
)

data class EdgeMeasuredRetrieval(
    val results: List<EdgeSearchResult>,
    val metrics: EdgeRetrievalMetrics
)

data class EdgeRAGMetrics(
    val retrieval: EdgeRetrievalMetrics,
    val generationMilliseconds: Double,
    val totalMilliseconds: Double,
    val requestedTopK: Int,
    val retainedResultCount: Int,
    val contextCharacterCount: Int
)

File: ios/Sources/EdgeFrameworks/Core/EdgeRAGMetrics.swift
public struct EdgeRetrievalMetrics: Sendable, Equatable {
    public let embeddingMilliseconds: Double
    public let searchMilliseconds: Double
    public let totalMilliseconds: Double
    public let resultCount: Int
    public let topScore: Float?
    public let bottomScore: Float?

    public init(
        embeddingMilliseconds: Double,
        searchMilliseconds: Double,
        totalMilliseconds: Double,
        resultCount: Int,
        topScore: Float?,
        bottomScore: Float?
    ) {
        self.embeddingMilliseconds = embeddingMilliseconds
        self.searchMilliseconds = searchMilliseconds
        self.totalMilliseconds = totalMilliseconds
        self.resultCount = resultCount
        self.topScore = topScore
        self.bottomScore = bottomScore
    }
}

public struct EdgeMeasuredRetrieval: Sendable, Equatable {
    public let results: [EdgeSearchResult]
    public let metrics: EdgeRetrievalMetrics

    public init(
        results: [EdgeSearchResult],
        metrics: EdgeRetrievalMetrics
    ) {
        self.results = results
        self.metrics = metrics
    }
}

public struct EdgeRAGMetrics: Sendable, Equatable {
    public let retrieval: EdgeRetrievalMetrics
    public let generationMilliseconds: Double
    public let totalMilliseconds: Double
    public let requestedTopK: Int
    public let retainedResultCount: Int
    public let contextCharacterCount: Int

    public init(
        retrieval: EdgeRetrievalMetrics,
        generationMilliseconds: Double,
        totalMilliseconds: Double,
        requestedTopK: Int,
        retainedResultCount: Int,
        contextCharacterCount: Int
    ) {
        self.retrieval = retrieval
        self.generationMilliseconds = generationMilliseconds
        self.totalMilliseconds = totalMilliseconds
        self.requestedTopK = requestedTopK
        self.retainedResultCount = retainedResultCount
        self.contextCharacterCount = contextCharacterCount
    }
}
